spring-web/src/main/java/org/springframework/web/multipart/reactive/nio/NioMultipartResolver.java

```diff
@@ -0,0 +1,167 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.web.multipart.reactive.nio;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import org.synchronoss.cloud.nio.multipart.Multipart;
+import org.synchronoss.cloud.nio.multipart.MultipartContext;
+import org.synchronoss.cloud.nio.multipart.MultipartUtils;
+import org.synchronoss.cloud.nio.multipart.NioMultipartParser;
+import org.synchronoss.cloud.nio.multipart.NioMultipartParserListener;
+import org.synchronoss.cloud.nio.stream.storage.StreamStorage;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.http.server.reactive.ServerHttpRequest;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.multipart.MultipartException;
+import org.springframework.web.multipart.reactive.Part;
+import org.springframework.web.multipart.reactive.MultipartResolver;
+
+/**
+ * {@link MultipartResolver} implementation based on the NIO Multipart library.
+ *
+ * The parsing of the request is non-blocking, but consuming the parts is not yet
+ * since NIO Multipart is currently using {@code InputStream} / {@code OutputStream}.
+ * See <a href="https://github.com/synchronoss/nio-multipart/issues/4">this GitHub issue</a>
+ * for more details about this limitation expected to be fixed in an upcoming release.
+ *
+ * @author Sebastien Deleuze
+ * @since 5.0
+ * @see <a href="https://github.com/synchronoss/nio-multipart">NIO Multipart</a>
+ */
+public class NioMultipartResolver implements MultipartResolver {
+
+	@Override
+	public boolean isMultipart(ServerHttpRequest request) {
+		if (!HttpMethod.POST.equals(request.getMethod())) {
+			return false;
+		}
+		MediaType contentType = request.getHeaders().getContentType();
+		return (contentType != null && MultipartUtils.isMultipart(contentType.toString()));
+	}
+
+	@Override
+	public Flux<Part> resolveParts(ServerHttpRequest request) {
+		if (!isMultipart(request)) {
+			return Flux.empty();
+		}
+		return Flux.create(new NioMultipartConsumer(request));
+	}
+
+	@Override
+	public Mono<MultiValueMap<String, Part>> resolveAllParts(ServerHttpRequest request) {
+		return resolveParts(request)
+				.collectList()
+				.map(parts -> {
+					MultiValueMap<String, Part> partsMap = new LinkedMultiValueMap<>();
+					parts.forEach(part -> partsMap.add(part.getName(), part));
+					return partsMap;
+				});
```

Comment:
Reactor `collectMultimap()` is using a `Collection` for the value type while `LinkedMultiValueMap` is taking a `Map<K, List<V>>` parameter, that's why I did not use it in my first implementation, but after more thoughts I was able to use it in my updated commit thanks to `Collectors.toMap()` to change the value type of the map.